queue/group_utils.go

```diff
@@ -0,0 +1,232 @@
+	Get(string) amboy.Queue
+	Remove(context.Context, string) error
+	Prune(context.Context) error
+	Close(context.Context) error
+	Names() []string
+	Len() int
+}
+
+// NewGroupCache produces a GroupCache implementation that supports a
+// default TTL setting, and supports cloning and closing operations.
+func NewGroupCache(ttl time.Duration) GroupCache {
+	if ttl == 0 {
+		ttl = time.Hour
+	}
+
+	return &cacheImpl{
+		ttl: ttl,
+		mu:  &sync.Mutex{},
+		q:   map[string]cacheItem{},
+	}
+}
+
+type cacheImpl struct {
+	ttl time.Duration
+	mu  *sync.Mutex
+	q   map[string]cacheItem
+}
+
+type cacheItem struct {
+	q    amboy.Queue
+	ttl  time.Duration
+	ts   time.Time
+	name string
+}
+
+func (c *cacheImpl) Len() int {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	return len(c.q)
+}
+
+func (c *cacheImpl) Names() []string {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	out := make([]string, len(c.q))
+	idx := 0
+	for name := range c.q {
+		out[idx] = name
+	}
+
+	return out
+}
+
+func (c *cacheImpl) Set(name string, q amboy.Queue, ttl time.Duration) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if _, ok := c.q[name]; ok {
+		return errors.Errorf("queue named '%s' is already cached", name)
+	}
+
+	if ttl == 0 {
+		ttl = c.ttl
+	}
+
+	c.q[name] = cacheItem{
+		q:    q,
+		ttl:  ttl,
+		ts:   time.Now(),
+		name: name,
+	}
+
+	return nil
+}
+
+func (c *cacheImpl) Get(name string) amboy.Queue {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if _, ok := c.q[name]; !ok {
+		return nil
+	}
+
+	item := c.q[name]
+	item.ts = time.Now()
+	c.q[name] = item
+
+	return item.q
+}
+
+func (c *cacheImpl) Remove(ctx context.Context, name string) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if _, ok := c.q[name]; !ok {
+		return nil
+	}
+
+	queue := c.q[name].q
+	if !queue.Stats().IsComplete() {
+		return errors.Errorf("cannot delete in progress queue, '%s'", name)
+	}
+
+	queue.Runner().Close(ctx)
+	delete(c.q, name)
+	return nil
+}
+
+func (c *cacheImpl) getCacheIterUnsafe() <-chan cacheItem {
+	out := make(chan cacheItem, len(c.q))
+	for _, v := range c.q {
+		out <- v
+	}
+	close(out)
+	return out
+}
+
+func (c *cacheImpl) getCacheIterSafe() <-chan cacheItem {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	return c.getCacheIterUnsafe()
+}
+
+func (c *cacheImpl) Prune(ctx context.Context) error {
+	wg := &sync.WaitGroup{}
+	num := runtime.NumCPU()
+	catcher := grip.NewBasicCatcher()
+	work := c.getCacheIterSafe()
+	wg.Add(num)
+
+	for i := 0; i < runtime.NumCPU(); i++ {
+		go func() {
+			defer recovery.LogStackTraceAndContinue("panic in local queue pruning")
+			defer wg.Done()
+			for {
+				select {
+				case <-ctx.Done():
+					catcher.Add(ctx.Err())
+					return
+				case item := <-work:
+					if item.ts.IsZero() {
+						return
+					}
+
+					if item.ttl > 0 && item.ttl > time.Since(item.ts) {
+						continue
+					}
+
+					if item.q.Stats().IsComplete() {
+						item.q.Runner().Close(ctx)
+
+						wait := make(chan struct{})
+						func() {
```

Comment:
Did you mean to put this in a goroutine?

Reply:
nope, want to be able to lock and use a defer but have it happen synchronously.

Reply:
jk, you're right.